### ISSUE TITLE
StatefulSet for VMAlertmanager will not be repaired once it goes into not-ready state

### ISSUE THREAD

**Issue by morimoto-cybozu** [VictoriaMetrics/operator#271]
## Description

Once I had misconfigured a VMAlertmanager CR and made the corresponding StatefulSet not-ready accidentally, I could not resume the StatefulSet by restoring the VMAlertmanager CR.
The only solution I could find was to delete the StatefulSet manually and to let the operator regenerate the StatefulSet.

## How to reproduce

1. Create a valid VMAlertmanager CR and wait for start-up of the corresponding StatefulSet and its Pods.
2. Edit the VMAlertmanager CR and introduce an error.  In my case, I misspelled a container image name for an additional container at `.spec.containers[0].image`.
3. Confirm that one of the Pods fails to start repeatedly and that Pod gets not-ready.  This is proper behavior.
4. Edit the VMAlertmanager CR to remove the error.
5. Wait for a while.
6. Confirm that the StatefulSet is not repaired.  It will still contain the misspelled container image name.

## Expected behavior

The StatefulSet is repaired to contain the correct container image name as written in the VMAlertmanager CR.  All the Pods are in the ready state.

## Related logs

The operator complains that it cannot start reconciliation because one of the already updated Pod is not ready.

```
{"level":"info","ts":1625482674.1232579,"logger":"controllers.VMAlertmanager","msg":"Reconciling","vmalertmanager":"monitoring/vmalertmanager-largeset"}
{"level":"info","ts":1625482680.7680185,"logger":"factory","msg":"sts update is needed","sts":"vmalertmanager-vmalertmanager-largeset","currentVersion":"vmalertmanager-vmalertmanager-largeset-84b5bf59f","desiredVersion":"vmalertmanager-vmalertmanager-largeset-854f766b8d"}
{"level":"info","ts":1625482680.7680783,"logger":"factory","msg":"checking if update needed","controller":"sts.rollingupdate","desiredVersion":"vmalertmanager-vmalertmanager-largeset-854f766b8d"}
{"level":"info","ts":1625482680.7685792,"logger":"factory","msg":"pod version doesnt match","controller":"sts.rollingupdate","desiredVersion":"vmalertmanager-vmalertmanager-largeset-854f766b8d","pod":"vmalertmanager-vmalertmanager-largeset-1","podVersion":"vmalertmanager-vmalertmanager-largeset-84b5bf59f"}
{"level":"info","ts":1625482680.768606,"logger":"factory","msg":"pod version doesnt match","controller":"sts.rollingupdate","desiredVersion":"vmalertmanager-vmalertmanager-largeset-854f766b8d","pod":"vmalertmanager-vmalertmanager-largeset-0","podVersion":"vmalertmanager-vmalertmanager-largeset-84b5bf59f"}
{"level":"info","ts":1625482680.76861,"logger":"factory","msg":"update is needed, start building proper order for update","controller":"sts.rollingupdate","desiredVersion":"vmalertmanager-vmalertmanager-largeset-854f766b8d"}
{"level":"info","ts":1625482680.7686353,"logger":"factory","msg":"updated pods with desired version:","controller":"sts.rollingupdate","desiredVersion":"vmalertmanager-vmalertmanager-largeset-854f766b8d","count":1}
{"level":"info","ts":1625482680.7686481,"logger":"factory","msg":"checking ready status for already updated pods to desired version","controller":"sts.rollingupdate","desiredVersion":"vmalertmanager-vmalertmanager-largeset-854f766b8d","pod":"vmalertmanager-vmalertmanager-largeset-2"}
{"level":"error","ts":1625482770.769468,"logger":"factory","msg":"cannot get ready status for already updated pod","controller":"sts.rollingupdate","desiredVersion":"vmalertmanager-vmalertmanager-largeset-854f766b8d","pod":"vmalertmanager-vmalertmanager-largeset-2","error":"timed out waiting for the condition","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tgithub.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/VictoriaMetrics/operator/controllers/factory.performRollingUpdateOnSts\n\tgithub.com/VictoriaMetrics/operator/controllers/factory/vmcluster.go:1489\ngithub.com/VictoriaMetrics/operator/controllers/factory.CreateOrUpdateAlertManager\n\tgithub.com/VictoriaMetrics/operator/controllers/factory/alertmanager.go:97\ngithub.com/VictoriaMetrics/operator/controllers.(*VMAlertmanagerReconciler).Reconcile\n\tgithub.com/VictoriaMetrics/operator/controllers/vmalertmanager_controller.go:83\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.8.1/pkg/internal/controller/controller.go:297\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.8.1/pkg/internal/controller/controller.go:252\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2\n\tsigs.k8s.io/controller-runtime@v0.8.1/pkg/internal/controller/controller.go:215\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99"}
{"level":"error","ts":1625482770.7699623,"logger":"controllers.VMAlertmanager","msg":"cannot create or update vmalertmanager sts","vmalertmanager":"monitoring/vmalertmanager-largeset","error":"cannot update statefulset for vmalertmanager: timed out waiting for the condition","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tgithub.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/VictoriaMetrics/operator/controllers.(*VMAlertmanagerReconciler).Reconcile\n\tgithub.com/VictoriaMetrics/operator/controllers/vmalertmanager_controller.go:85\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.8.1/pkg/internal/controller/controller.go:297\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.8.1/pkg/internal/controller/controller.go:252\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2\n\tsigs.k8s.io/controller-runtime@v0.8.1/pkg/internal/controller/controller.go:215\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99"}
{"level":"error","ts":1625482770.7700963,"logger":"controller-runtime.manager.controller.vmalertmanager","msg":"Reconciler error","reconciler group":"operator.victoriametrics.com","reconciler kind":"VMAlertmanager","name":"vmalertmanager-largeset","namespace":"monitoring","error":"cannot update statefulset for vmalertmanager: timed out waiting for the condition","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tgithub.com/go-logr/zapr@v0.2.0/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.8.1/pkg/internal/controller/controller.go:301\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.8.1/pkg/internal/controller/controller.go:252\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2\n\tsigs.k8s.io/controller-runtime@v0.8.1/pkg/internal/controller/controller.go:215\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\tk8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99"}
```

## Suspicious code

The usage of `performRollingUpdateOnSts()` in `controllers/factory/alertmanager.go` seems inappropriate as compared with the usage in `controllers/factory/vmcluster.go`.
In `vmcluster.go`, the caller first updates the StatefulSet on the API server and then calls `performRollingUpdateOnSts()`.  The goal of the rolling update is set to the new StatefulSet.
In `alertmanager.go`, the caller calls `performRollingUpdateOnSts()` without updating the StatefulSet.  The goal is not renewed.  The operator misunderstands that the failing Pod is in the latest version.

**Comment by f41gh7** [comment]
Hello, thanks for reporting and investigating, it will be fixed soon.

**Comment by f41gh7** [comment]
Changes was included to the 0.16.0 release, feel free to re-open, if problem persists.
